### PR TITLE
fix: pq cosine fast lookup table

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/lance/src/index/vector/pq.rs [hunk 1]
```diff
@@ -130,65 +129,63 @@ impl PQIndex {
...
             let y_norm = sub_vector_centroids
-                .as_ref()
                 .values()
                 .chunks_exact(sub_vector_length)
-                .map(|cent| {
-                    // Accelerate this later.
-                    cent.iter().map(|y| y.powi(2)).sum::<f32>()
-                });
-            y_norm_table.extend(y_norm);
+                .map(|cent| cent.norm_l2().powf(2.0));
```
Comment (westonpace) [inline rust/lance/src/index/vector/pq.rs:165]: Can you drop the `.powf(2.0)` here and the `.sqrt` of `y_norm.sqrt()` below?  Not a correctness thing, just a minor perf suggestion.

**Review by westonpace — APPROVED** [review]
This matches my understanding of cosine distance.

File: rust/lance/src/index/vector/pq.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/pq.rs:165]: We can not. 

so this is an array of `y^2`.    

$`\sum{\sqrt{y^2}} != \sum_{0}^{n}{\sqrt{y_i^2}}`$

Comment (westonpace) [reply rust/lance/src/index/vector/pq.rs:165]: It just seems like the two operations will cancel each other out. If you remove all the middle code you basically have...

```
let y_norm_for_cent = cent.norm_l2().powf(2.0);
let cosine_dist = `1.0 - xy / (x_norm_for_centroid * y_norm_for_centroid.sqrt())
```

So the right term is basically: `cent.norm_l2().powf(2.0).sqrt()` which should be identical to `cent.norm_l2()` right?

Comment (eddyxu) [reply rust/lance/src/index/vector/pq.rs:165]: y_norm_table build a table of y^2 for each sub-vector. 
the last `y_norm_for_centroid` is the sum of these y^2 for each PQ subvector tho.

**Pull request merged**